qualtran/bloqs/state_preparation/state_preparation_via_rotation.py

```diff
@@ -0,0 +1,443 @@
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    selection_bitsizes: int
+    rot_reg_bitsizes: int
+    rom_values: Tuple[int, ...]
+    control_bitsize: int
+
+    @property
+    def signature(self) -> Signature:
+        return Signature.build(
+            prepare_control=self.control_bitsize,
+            selection=self.selection_bitsizes,
+            qubit=1,
+            rot_reg=self.rot_reg_bitsizes,
+            phase_gradient=self.rot_reg_bitsizes,
+        )
+
+    def build_composite_bloq(self, bb: BloqBuilder, **soqs: SoquetT) -> Dict[str, SoquetT]:
+        """Parameters:
+        * prepare_control (not necessary if selection_bitsizes == 0)
+        * selection (not necessary if selection_bitsizes == 0)
+        * qubit
+        * rot_reg
+        * phase_gradient
+        """
+        qrom = QROM(
+            [np.array(self.rom_values)],
+            selection_bitsizes=(self.selection_bitsizes,),
+            target_bitsizes=(self.rot_reg_bitsizes,),
+            num_controls=self.control_bitsize + 1,
+        )
+        if self.control_bitsize == 0:
+            qrom_control = soqs.pop("qubit")
+        else:
+            # both prepare_control and qubit will control the QROM so that this acts as a control Z
+            qrom_control = bb.join(
+                np.array([*bb.split(soqs.pop("prepare_control")), soqs.pop("qubit")])
+            )
+        phase_grad = soqs.pop("phase_gradient")
+        # load angles in rot_reg (line 1 of eq (8) in [1])
+        soqs = bb.add_d(qrom, control=qrom_control, target0_=soqs.pop("rot_reg"), **soqs)
+        # phase kickback via phase_grad += rot_reg (line 2 of eq (8) in [1])
+        soqs["target0_"], phase_grad = bb.add(
+            AddIntoPhaseGrad(self.rot_reg_bitsizes, self.rot_reg_bitsizes),
+            x=soqs["target0_"],
+            phase_grad=phase_grad,
+        )
+        # uncompute angle load in rot_reg to disentangle it from selection register
+        # (line 1 of eq (8) in [1])
+        soqs = bb.add_d(qrom, **soqs)
+        separated = bb.split(soqs.pop('control'))
+        if self.control_bitsize != 0:
+            soqs["prepare_control"] = bb.join(separated[:-1])
+        return {
+            "qubit": separated[-1],
+            "phase_gradient": phase_grad,
+            "rot_reg": soqs.pop('target0_'),
+        } | soqs
+
+
+class RotationTree:
+    r"""Used by `StatePreparationUsingRotations` to get the corresponding rotation angles.
+
+    The rotation angles are used to encode the amplitude of a state using the method described in
+    [1], section on arbitrary quantum state preparation, page 3.
+
+    The only methods to be used externally are extract_rom_values_from_state, angle_to_rom_value and
+    build_rotation_tree_from_state.
```

Comment:
the names of the methods in the docstrings don't match the new names of the methods